gammagl/layers/conv/hgt_conv.py

```diff
@@ -0,0 +1,151 @@
+        for node_type, x in x_dict.items():
+            k_dict[node_type] = tlx.ops.reshape(self.k_lin[node_type](x), (-1, H, D))
+            q_dict[node_type] = tlx.ops.reshape(self.q_lin[node_type](x), (-1, H, D))
+            v_dict[node_type] = tlx.ops.reshape(self.v_lin[node_type](x), (-1, H, D))
+            out_dict[node_type] = []
+
+        # Iterate over edge-types:
+        for edge_type, edge_index in edge_index_dict.items():
+            transpose = tlx.nn.Transpose([1, 0, 2])
+            src_type, _, dst_type = edge_type
+            edge_type = '__'.join(edge_type)
+
+            a_rel = self.a_rel[edge_type + 'a']
+            k = transpose((transpose(k_dict[src_type]) @ a_rel))
+
+            m_rel = self.m_rel[edge_type + 'm']
+            v = transpose((transpose(v_dict[src_type]) @ m_rel))
+
+            if tlx.BACKEND != 'tensorflow':
+                edge_index = tlx.ops.convert_to_tensor(edge_index, dtype='int64')
+            source_index, target_index = edge_index[0], edge_index[1]
+            q_i = tlx.gather(q_dict[dst_type], target_index, axis=0)
+            v_j = tlx.gather(v, source_index, axis=0)
+            k_j = tlx.gather(k, source_index, axis=0)
+            rel = self.p_rel[edge_type]
+            out = self.propagate(edge_index=edge_index, aggr='sum', q_i=q_i, k_j=k_j, v_j=v_j, rel=rel,
+                                 num_nodes=x_dict[dst_type].shape[0])
+            out_dict[dst_type].append(out)
+
+        # Iterate over node-types:
+        for node_type, outs in out_dict.items():
+            outs = tlx.stack(outs)
+            out = tlx.ops.reduce_sum(outs, axis=0, keepdims=False)
+            out = self.a_lin[node_type](out)
+            alpha = tlx.ops.sigmoid(self.skip[node_type])
+            out = alpha * out + (1 - alpha) * x_dict[node_type]
+            out_dict[node_type] = out
+
+        return out_dict
+
+    def propagate(self, edge_index, aggr='sum', **kwargs):
+        coll_dict = {}
+        for k, v in kwargs.items():
+            coll_dict[k] = v
+        coll_dict['edge_index'] = edge_index
+        coll_dict['aggr'] = aggr
+        coll_dict['target_index'] = edge_index[1]
+        msg_kwargs = self.inspector.distribute('message', coll_dict)
+        msg = self.message(**msg_kwargs)
+        x = self.aggregate(msg, edge_index, num_nodes=kwargs['num_nodes'], aggr=aggr)
+        x = self.update(x)
+        return x
+
+    def message(self, k_j, q_i, v_j, rel, target_index, num_nodes):
+        alpha = tlx.ops.reduce_sum(k_j * q_i, axis=-1, keepdims=False)
+        alpha = alpha * rel
+        alpha = alpha / math.sqrt(q_i.shape[-1])
+        alpha = self.dropout(segment_softmax(alpha, target_index, num_nodes))
+        out = v_j * tlx.expand_dims(alpha, -1)
+        return tlx.ops.reshape(out, (-1, self.out_channels))
+
+    def aggregate(self, msg, edge_index, num_nodes=None, aggr='sum'):
+        return super().aggregate(msg, edge_index, num_nodes, aggr)
```

Comment:
I think we do not need this function. In "self.propagate", we will automatically call the aggregate function in "gammagl.layers.conv.message_passing" if you do not overwrite it. So, I suggest we should delete this function for coding simplicity.